include/ignition/math/Capsule.hh

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_MATH_CAPSULE_HH_
+#define IGNITION_MATH_CAPSULE_HH_
+
+#include "ignition/math/MassMatrix3.hh"
+#include "ignition/math/Material.hh"
+
+namespace ignition
+{
+  namespace math
+  {
+    // Foward declarations
+    class CapsulePrivate;
+
+    // Inline bracket to help doxygen filtering.
+    inline namespace IGNITION_MATH_VERSION_NAMESPACE {
+    //
+    /// \class Capsule Capsule.hh ignition/math/Capsule.hh
+    /// \brief A representation of a capsule or sphere-capped cylinder.
+    ///
+    /// The capsule class supports defining a capsule with a radius,
+    /// length, and material properties. The shape is equivalent to a cylinder
+    /// aligned with the Z-axis and capped with hemispheres. Radius and
+    /// length are in meters. See Material for more on material properties.
+    template<typename Precision>
+    class Capsule
+    {
+      /// \brief Default constructor. The default radius and length are both
+      /// zero.
+      public: Capsule() = default;
```

Comment:
What do you think about not providing a default constructor? That way the only time we would have invalid Capsule objects would be if the input length and radius values are invalid.

Reply:
I like having the default constructor. If you want to create a shape without remembering the order of the arguments in the constructor:

~~~
Capsuled capsule;
capsule.SetLength(length);
capsule.SetRadius(radius);
~~~

if other people don't like it, we can remove it, but I'm inclined to keep it

Reply:
If I put on my pedantry hat, the "right" way to do what @scpeters is suggesting would be to have something like a `Capsule::Builder` class that lets you fill in the information necessary to build a `Capsule` object, and then you could call `build()` on it (or something similar) and it constructs a capsule for you, perhaps throwing an exception if the parameters are invalid (e.g. radius <= 0, length <= 0).

Some would (fairly) argue that's overkill for this use case, but the more I've had to deal with writing production-quality code, the more I've been feeling it's worth the effort to make every assumption explicit inside the code itself.